### PR TITLE
third layer of vpn check via ipqs ENG-1789

### PR DESCRIPTION
**Pull request by sanandnarayan**
# Task: 
VPN - implement the third layer of blocking for opyn.co
## Description

VPN - implement the third layer of blocking for opyn.co
Fixes (linear-task) ENG-1789

## Type of change

- [x] New feature
- [ ] Bug fix
- [ ] Testing code
- [ ] Document update or config files

### CODE REVIEW

**Comment by sanandnarayan** [comment]
before merging: Do check the redis tokens permission with K

File: packages/frontend/src/server/ipqs.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  bot_status: boolean
+  connection_type: string
+  abuse_velocity: string
+  zip_code: string
+  latitude: number
+  longitude: number
+  request_id: string
+}
+
+async function CheckUserIP(ip_address: string, user_agent: string, language: string): Promise<ApiResponse | null> {
```
Comment (iamoperand) [inline packages/frontend/src/server/ipqs.ts:34]: i prefer using camelCase for function and variable names
```suggestion
async function checkUserIP(ipAddress: string, userAgent: string, language: string): Promise<ApiResponse | null> {
```

File: packages/frontend/src/server/ipqs.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    allow_public_access_points
+  const result = await get_IPQ_URL(url)
+  if (result !== null) {
+    return result
+  } else {
+    return null
+  }
+}
+
+async function get_IPQ_URL(url: string): Promise<ApiResponse | null> {
```
Comment (iamoperand) [inline packages/frontend/src/server/ipqs.ts:55]: this is just a fetcher function right? name didn't make sense to me

File: packages/frontend/src/server/ipqs.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  }
+}
+
+export async function isVPN(ip_address: string, user_agent: string, language: string): Promise<boolean> {
+  const ip_result = await CheckUserIP(ip_address, user_agent, language)
+  if (ip_result === null) {
+    return false
+  }
+
+  if (typeof ip_result['vpn'] !== 'undefined' && ip_result['vpn'] === true) {
```
Comment (iamoperand) [inline packages/frontend/src/server/ipqs.ts:71]: can simplify this to:
`return (ip_result['vpn'] === true) || (ip_result['tor'] === true);`

for things like this, i tend to use sidekick. just say, help me simplify this piece of code

File: packages/frontend/src/server/ipqs.ts (hunk 1 above)
Comment (iamoperand) [reply packages/frontend/src/server/ipqs.ts:34]: also `user_agent` and `language` unused

**Comment by sanandnarayan** [comment]
@iamoperand appreciate the comments. From a logic perspective does this look good? styling I can fix seperately later if required

**Comment by iamoperand** [comment]
> @iamoperand appreciate the comments. From a logic perspective does this look good? styling I can fix seperately later if required

yeah looks good logic wise. tested it through nordvpn (australia server) and it worked.

**Review by iamoperand — APPROVED** [review]

File: packages/frontend/middleware.ts [hunk 4]
```diff
@@ -11,7 +12,9 @@ export async function middleware(request: NextRequest) {
   const country = cloudflareCountry ?? request.geo?.country
   const url = request.nextUrl
 
-  const ip = request.headers.get('cf-connecting-ip')
+  const ip = request.headers.get('cf-connecting-ip') || request.headers.get('x-forwarded-for') || request.ip
+  const user_agent = request.headers.get('user-agent')
```
Comment (alpinechicken) [inline packages/frontend/middleware.ts:16]: why are we getting `user_agent` and `language`?

**Review by alpinechicken — APPROVED** [review]
lgtm

- should change blocked message so it's clear that it's either a vpn or a blocked country

**Pull request merged**